### ISSUE TITLE
support settings to ignore rules for glob files

### ISSUE THREAD

**Issue by tjx666** [DavidAnson/markdownlint#666]
In my large monorepo projects, there are many changelog file, I want to disable MD041 for all changelog file. For now, I can't find a simple way to achieve this.

**Comment by DavidAnson** [comment]
Assuming you can identify changelog when invoking the linter, what about running two passes for linting? One pass with that rule disabled and only looking at changelogs, the other pass scanning everything else.

**Comment by DavidAnson** [comment]
(This library doesn't do anything with globs, it just scans the list of files you give it.)

**Comment by tjx666** [comment]
I config the markdownlint using `.markdownlint.json`, any way to satisfy my need only modify the config?

My suggestion is to support something like eslint overrides option.

```json
{
    "overrides" {
        "files": ["CHANGELOG.md"],
        "rules": {
            ""MD001": false,
        }
    }
}

```

**Comment by nschonni** [comment]
Think it's similar to https://github.com/DavidAnson/markdownlint-cli2/issues/50 and is looking for the support that other libraries like ESLint https://eslint.org/docs/latest/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns or Prettier https://prettier.io/docs/en/configuration.html#configuration-overrides

**Comment by stern-shawn** [comment]
> OK, thanks. I will leave this issue here as a reminder to myself, but this may end up making the most sense in CLI or CLI2.

Hi @DavidAnson , is this still on the radar? I've run into a scenario where we have `.mdx` and `.snippet.mdx` files in a project, and would like to disable _some_ rules such as [MD041](https://github.com/DavidAnson/markdownlint/blob/main/doc/md041.md) only for the reusable snippet files, but otherwise apply the remaining rules to both types. Something such as the `overrides` syntax would be wonderful for this, as opposed to doing two separate runs with different configs.

Ex.
```json
{
  "overrides" {
    "files": ["**/*.snippet.mdx"],
    "rules": {
      "MD041": false,
    }
  }
}
```

**Comment by DavidAnson** [comment]
@stern-shawn, when using this library directly, you can pass it whatever set of files you want and should filter out any you don't. When using markdownlint-cli or -cli2, you may be able to use the ignore argument or option to filter out unwanted files. That may not work for all scenarios, but it's hard to say in the abstract.

**Comment by stern-shawn** [comment]
@DavidAnson I see, and my apologies (I found this via google search and didn't confirm that it was part of the markdownlint-cli2 issues 🤦 ). I will look at running -cli2 with one pass for snippet files and another for all other mdx 🙇

**Comment by DavidAnson** [comment]
Closing the issue for the library because the new `overrides` property of `.markdownlint-cli2.jsonc` (and similar) will support the desired behavior in the next release of that package. (I decided this behavior belonged in CLI2 and not the library: <https://github.com/DavidAnson/markdownlint-cli2/issues/797>).

**Comment by beporter** [comment]
I'm not familiar with the broader architecture, but doesn't restricting the ability to specify file overrides to only the cli2 tool mean that things like the VScode extension can't take advantage of it?

The critical characteristic for me is that the config file needs to be authoritative for the entire git repository it resides in. It's a non-starter to have to specify a list of rule-exclusions per file (or glob), per implementation. That should be what the config file is for, after all.

## Tangent Territory

There's lots of prior art for this too. PHP_CodeSniffer for example [uses klunky xml](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/4.x/phpcs.xml.dist), but provides really granular abilities to scope rules to specific files as well as adjust their priority.

From their [annotated rule file](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Annotated-Ruleset):

```xml
  <rule ref="Squiz.Strings.DoubleQuoteUsage"> <!-- For this specific rule... -->
    <type>warning</type> <!-- ...override from error down to warning, -->

    <severity>3</severity> <!-- ...and reduce the severity down to 3, -->

    <!-- ...but only for test-related folders! -->
    <include-pattern>*/tests/*</include-pattern>
    <include-pattern>*/data/*</include-pattern>

    <!-- This rule remains an error in the rest of the code base. -->
 </rule>
```

I get the desire for a separation of concerns, but it's _the library_ that's ultimately interacting with the files, not solely the cli tooling, isn't it?

**Comment by DavidAnson** [comment]
@beporter

> ... doesn't restricting the ability to specify file overrides to only the cli2 tool mean that things like the VScode extension can't take advantage of it?

No. The VS Code extension is documented to support the same configuration files and behavior as CLI2. (It does this by running the same code, though that's an implementation detail.)

https://github.com/DavidAnson/vscode-markdownlint#configure
